### ISSUE TITLE
Send code instead of magic link

### ISSUE THREAD

**Issue by vbylen** [supabase/gotrue#264]
# Feature request

## Is your feature request related to a problem? Please describe.

On mobile, the current magic link implementation requires the user to manually navigate to the mail app, find the email (in their spam folder) and click the link.

Along the way the OS may have presented the user with an option to choose a specific mail app, and/or to choose which browser to use to open the link.

There's also the developer headache of implementing redirects which can get annoying when using local, staging and production environments.

All of the above adds cognitive overload.

**tl;dr** sending a short code to the user will improve UX and DX.

## Describe the solution you'd like

Instead of sending a magic link, send a short code.

No need for context switching between apps (assuming the user has email notifications enabled).

## Additional context

Like so:
![image](https://user-images.githubusercontent.com/57633368/128057314-feb46191-2e68-48b7-866c-6287036e7fee.jpeg)

**Comment by edgarasben** [comment]
Described this issue also at [#4455](https://github.com/supabase/supabase/issues/4455)

**Comment by edgarasben** [comment]
What I would add to this, please don't make it code OR link. There should be an option to send code AND link for user to choose from. I talked with some people about this and for some it's a personal preference.

**Comment by awalias** [comment]
we also need to make sure to reference https://github.com/supabase/gotrue/issues/290 some valuable discussion there

**Comment by liaujianjie** [comment]
I'd also like to add that some of our users' work email is configured to sandbox or sanitise URLs found in emails. For example, with [Mimecast](https://www.mimecast.com/products/email-security-with-targeted-threat-protection/url-protect/). Not sure why but magic link URLs, when opened in URL sandboxes, seem to not sign in.

**Comment by liaujianjie** [comment]
Is it possible to bump this up in the list of priorities? I'm receiving user support requests regarding this sign in issue weekly, so I'll be the first in line to test this when shipped. 😅

Use cases where this feature is important for us:
1. When the user receives the magic link email on a different device or browser from which they are trying to sign in from.
2. When the user's work email has URL sanitization/sandboxing.

**Comment by J0** [comment]
Done. Let's talk about this tomorrow.

**Comment by LeafBlader123** [comment]
Can we get this in? it's bizarre to me this was not incorporated when designing the user API's. If on a shared computer, or if a hacker gets a token credentials, the attack blast radius is much larger. They can literally change the user's password and get full access to the account

**Comment by paintingStyle** [comment]
我也想要验证码登录的功能，能加入这个功能吗？

**Comment by edgarasben** [comment]
Another example from Notion.so how they do it:
<img width="448" alt="image" src="https://user-images.githubusercontent.com/725324/175818563-15700d56-6511-43d5-8a60-d37efae9c080.png">

**Comment by liaujianjie** [comment]
@edgarasben It is already possible :-)

In your email templates:

```
<p>
  Use the login code <code>{{ .Token }}</code>.
</p>
```

On the client side:

```tsx
await auth.verifyOTP({ email: "janedoe@compant.com", token: "OTP_TOKEN_GOES_HERE", type: "magiclink" });
```

**Comment by J0** [comment]
Thanks for highlighting this @liaujianjie !

Going to go ahead and close the issue but feel free to re-open if anyone still has related requests/unresolved concerns.

**Comment by vbylen** [comment]
@liaujianjie any way to test this in a local environment with supabase cli? I can view the templates tab but the ui won't let me edit them.

**Comment by LianBond** [comment]
How do you confirm the OTP with bash?

**Comment by ahmadbk** [comment]
I am not sure if the response by [liaujianjie](https://github.com/liaujianjie) resolves this. How do you send the Otp given just an email? The idea was to allow users to verify email address using Otp over Magic Link during the signup process.

Am I missing something?

**Comment by kangmingtay** [comment]
@ahmadbk once you've added `{{ .Token }}` to your email templates, the email sent to the user will contain a numeric code for them to enter

you can call the [`verifyOtp`](https://supabase.com/docs/reference/javascript/auth-verifyotp) method to verify that code - have you tried this out yet?
